evm-ds/src/precompiles/scilla_read.rs

```diff
@@ -0,0 +1,252 @@
+
+    let chunks = input_field_type.split_whitespace().collect::<Vec<_>>();
+    if chunks.len() == 1 {
+        return Ok(ScillaField {
+            indices: vec![],
+            ret_type: chunks[0].to_string(),
+        });
+    }
+    // We support only maps and maps of maps
+    let mut idx = 0;
+    let mut field_types = vec![ParamType::Address, ParamType::String];
+    while idx < chunks.len() {
+        if idx == chunks.len() - 1 {
+            break;
+        }
+        let chunk = chunks[idx];
+        if chunk.eq("Map") {
+            if idx + 2 >= chunks.len() {
+                // Malformatted json
+                return Err(PrecompileFailure::Error {
+                    exit_status: ExitError::Other(Cow::Borrowed(
+                        "Unable to find field with given name in fields list",
+                    )),
+                });
+            }
+            field_types.push(substitute_scilla_type_with_sol(chunks[idx + 1])?);
+            idx += 2;
+        } else {
+            field_types.push(substitute_scilla_type_with_sol(chunk)?);
+            idx += 1;
+        }
+    }
+
+    let Ok(decoded_array) = decode(&field_types, input) else {
+        return Err(PrecompileFailure::Error {
+            exit_status: ExitError::Other(Cow::Borrowed("Unable to decode input by specified field definition")),
+        });
+    };
+
+    let mut indices = vec![];
+    for value in decoded_array.iter().skip(2) {
+        match value {
+            Token::Uint(solidity_uint) => {
+                indices.push(format!("{}", solidity_uint));
+            }
+            Token::Address(solidity_addr) => {
+                indices.push(format!("0x{}", hex::encode(solidity_addr)));
+            }
+            _ => {
+                indices.push(value.to_string());
+            }
+        }
+    }
+    Ok(ScillaField {
+        indices,
+        ret_type: chunks.last().unwrap().to_string(),
+    })
+}
+
+fn extract_substate_from_json(
+    value: &Value,
+    vname: &str,
+    def: &ScillaField,
+) -> Result<Option<Token>, PrecompileFailure> {
+    if !value.is_object() || value.get(vname).is_none() {
+        return Ok(None);
+    }
+    if def.indices.is_empty() {
+        return encode_result_type(&value[vname], def);
+    }
+    let mut value = value[vname].as_object().unwrap();
+
+    for index in def.indices.clone().iter().take(def.indices.len() - 1) {
+        if !value[index].is_object() {
+            return Err(PrecompileFailure::Error {
+                exit_status: ExitError::Other(Cow::Borrowed(
+                    "Scilla field definition doesn't match with returned value",
+                )),
+            });
+        }
+        value = value[index].as_object().unwrap();
+    }
+
+    encode_result_type(&value[def.indices.last().unwrap()], def)
+}
+
+fn encode_result_type(
+    value: &Value,
+    def: &ScillaField,
+) -> Result<Option<Token>, PrecompileFailure> {
+    if def.ret_type.starts_with("Uint") {
```

Comment:
I'm not sure I'm entirely happy with the amount of information being destroyed here - though I guess there is no way to return a null value ("something went wrong decoding this and I don't know what the value is")? In particular, defining null to be 0 for ints is not a great idea if you can avoid it?